File: backend/src/models/Usuario.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');
const bcrypt = require('bcryptjs');

const Usuario = sequelize.define('Usuario', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    nome: {
        type: DataTypes.STRING,
        allowNull: false
    },
    email: {
        type: DataTypes.STRING,
        allowNull: false,
        unique: true,
        validate: {
            isEmail: true
        }
    },
    senha: {
        type: DataTypes.STRING,
        allowNull: false
    },
    telefone: {
        type: DataTypes.STRING
    },
    cpf: {
        type: DataTypes.STRING,
        unique: true
    },
    tipo: {
        type: DataTypes.ENUM('cidadao', 'equipe', 'gestor', 'admin'),
        defaultValue: 'cidadao'
    },
    ativo: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    },
    avatar: {
        type: DataTypes.STRING
    }
}, {
    tableName: 'usuarios',
    hooks: {
        beforeCreate: async (usuario) => {
            if (usuario.senha) {
                usuario.senha = await bcrypt.hash(usuario.senha, 10);
            }
        },
        beforeUpdate: async (usuario) => {
            if (usuario.changed('senha')) {
                usuario.senha = await bcrypt.hash(usuario.senha, 10);
            }
        }
    }
});

// Método para verificar senha
Usuario.prototype.verificarSenha = async function(senha) {
    return await bcrypt.compare(senha, this.senha);
};

// Método para serializar (não retornar senha)
Usuario.prototype.toJSON = function() {
    const values = Object.assign({}, this.get());
    delete values.senha;
    return values;
};

module.exports = Usuario;

File: python-zeladoria/app/models/usuario.py
from sqlalchemy import Column, Integer, String, Boolean, ForeignKey
from sqlalchemy.orm import relationship
from app.database.database import Base
import bcrypt


class Usuario(Base):
    __tablename__ = "usuarios"

    id = Column(Integer, primary_key=True, index=True)
    nome = Column(String, nullable=False)
    email = Column(String, unique=True, index=True, nullable=False)
    senha = Column(String, nullable=False)
    telefone = Column(String)
    cpf = Column(String, unique=True)
    tipo = Column(String, default="cidadao")  # cidadao, equipe, gestor, secretaria, admin
    ativo = Column(Boolean, default=True)
    avatar = Column(String)
    
    # Relacionamento com Secretaria (NULLABLE)
    secretaria_id = Column(Integer, nullable=True)

    # Relacionamentos
    chamados = relationship("Chamado", back_populates="usuario", foreign_keys="Chamado.usuario_id")
    chamados_responsavel = relationship("Chamado", back_populates="responsavel", foreign_keys="Chamado.responsavel_id")
    comentarios = relationship("Comentario", back_populates="usuario")

    def verificar_senha(self, senha: str) -> bool:
        """Verifica senha usando bcrypt diretamente"""
        try:
            # Truncar para 72 bytes
            senha_bytes = senha.encode('utf-8')[:72]
            senha_hash = self.senha.encode('utf-8')
            return bcrypt.checkpw(senha_bytes, senha_hash)
        except Exception as e:
            print(f"Erro ao verificar senha: {e}")
            return False

    @staticmethod
    def hash_senha(senha: str) -> str:
        """Hash senha usando bcrypt diretamente"""
        # Truncar para 72 bytes
        senha_bytes = senha.encode('utf-8')[:72]
        # Gerar hash
        hashed = bcrypt.hashpw(senha_bytes, bcrypt.gensalt())
        # Retornar como string
        return hashed.decode('utf-8')
